src/keyboard-lomiri.c

```diff
@@ -232,15 +198,56 @@ void keyboard_GetLayout(Keyboard *pKeyboard, gint nLayout, gchar **pLanguage, gc
     {
         *pDescription = g_strdup(pLayout->sDescription);
     }
-
-    g_free(sId);
 }
 
 void keyboard_SetLayout(Keyboard *pKeyboard, gint nLayout)
 {
     if (isGreeter() == FALSE)
     {
-        // TODO
+        gchar *sId = g_slist_nth_data (pKeyboard->pPrivate->lLayoutRec, nLayout);
+
+        GVariantBuilder cSettingsBuilder;
+        g_variant_builder_init (&cSettingsBuilder, G_VARIANT_TYPE ("a(ss)"));
+        g_variant_builder_add (&cSettingsBuilder, "(ss)", "xkb", sId);
+
+        GVariantBuilder cAccountsBuilder;
+        g_variant_builder_init (&cAccountsBuilder, G_VARIANT_TYPE ("aa{ss}"));
+        GVariantBuilder cDictBuilder;
+        g_variant_builder_init (&cDictBuilder, G_VARIANT_TYPE ("a{ss}"));
+        g_variant_builder_add (&cDictBuilder, "{ss}", "xkb", sId);
+        GVariant *pDict = g_variant_builder_end (&cDictBuilder);
+        g_variant_builder_add_value (&cAccountsBuilder, pDict);
+
+        guint nSources = g_slist_length (pKeyboard->pPrivate->lLayoutRec);
+
+        for (guint nSource = 0; nSource < nSources; nSource++)
+        {
+            if (nSource != nLayout)
+            {
+                gchar *sId = g_slist_nth_data (pKeyboard->pPrivate->lLayoutRec, nSource);
+
+                g_variant_builder_add (&cSettingsBuilder, "(ss)", "xkb", sId);
+
+                GVariantBuilder cDictBuilder;
+                g_variant_builder_init (&cDictBuilder, G_VARIANT_TYPE ("a{ss}"));
+                g_variant_builder_add (&cDictBuilder, "{ss}", "xkb", sId);
+                GVariant *pDict = g_variant_builder_end (&cDictBuilder);
+                g_variant_builder_add_value (&cAccountsBuilder, pDict);
+            }
+        }
+
+        GVariant *pSettingsSources = g_variant_builder_end (&cSettingsBuilder);
+        g_settings_set_value (pKeyboard->pPrivate->pSettings, "sources", pSettingsSources);
+
+        GDBusConnection *pConnection = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, NULL);
+        gint nUid = geteuid ();
+        gchar *sPath = g_strdup_printf ("/org/freedesktop/Accounts/User%i", nUid);
+        GDBusProxy *pProxy = g_dbus_proxy_new_sync (pConnection, G_DBUS_PROXY_FLAGS_NONE, NULL, "org.freedesktop.Accounts", sPath, "org.freedesktop.DBus.Properties", NULL, NULL);
+        g_free (sPath);
+        GVariant *pAccountsSources = g_variant_builder_end (&cAccountsBuilder);
+        GVariant *pRet = g_dbus_proxy_call_sync (pProxy, "Set", g_variant_new ("(ssv)", "com.lomiri.shell.AccountsService", "InputSources", pAccountsSources), G_DBUS_CALL_FLAGS_NONE, -1, NULL, NULL);
```

Comment:
for Lomiri on Ubuntu, we need UBUNTU_ACCOUNTSSERVICE_ENABLED here, don't we? Otherwise this will not work in Lomiri (Greeter) on Ubuntu, right?

Reply:
Added UBUNTU_ACCOUNTSSERVICE_ENABLED option.